DesignSafe-CI/portal
# feat: WP-664 `external_href` for app variant

## Overview / Changes

- Add a `download_href` for app variant.
- Move required fields out of "Tapis …" fieldset.
- Distinguish `href` and `download_href`.

## PR

* Work in Progress.
* Hold.

## Related

* [<del>WA-326</del> <ins>WP-664</ins>](https://tacc-main.atlassian.net/browse/WP-664)

## Testing

1. Enter CMS container.
    <sup>`docker exec -it des_django bash`</sup>
2. [Run migration.](https://docs.djangoproject.com/en/5.0/ref/django-admin/#migrate)
    <sup>`./manage.py migrate workspace 0018`</sup>
1. Create/Have a CMS page that lists an enabled app variant.
2. Save **Download href** to a path or URL e.g 
    <sup>`/data/browser/tapis/designsafe.storage.community/%2FSimCenter%2FSoftware%2FEE_UQ`</sup>[^1]
3. Load the page.
4. **Verify** variant has enabled "Download" button.
5. **Verify** "Download" button opens path/URL from step 2.
6. [Undo migration.](https://docs.djangoproject.com/en/5.0/ref/django-admin/#migrate)
    `./manage.py migrate workspace 0017`</sup>

[^1]: Source of this URL is "Download App" button on https://simcenter.designsafe-ci.org/research-tools/ee-uq-application/. ⚠️ I failed to load `/data/browser/` locally, but I think this is my environment, not my code.

## UI

<details open><summary>

### New Feature

</summary>

#### Has External HREF, No Version

| render | data `href` | data `external_href` |
| - | - | - |
| <img width="1024" alt="external - render" src="https://github.com/user-attachments/assets/48851596-2764-41c0-8ff8-083e101ba0a3"> | <img width="1024" alt="external - href" src="https://github.com/user-attachments/assets/2c67154a-73c5-4eec-9d11-60f1d30aa07b"> | <img width="1024" alt="external - external_href" src="https://github.com/user-attachments/assets/5149ad6d-19ba-403a-b8f5-aa602fdd7080"> |

#### No External HREF, Has Version

| render | data `href` | data `external_href` |
| - | - | - |
| <img width="1024" alt="has version - render" src="https://github.com/user-attachments/assets/81db4fb6-7102-447c-9008-19a8f615d450"> | <img width="1024" alt="has version - href" src="https://github.com/user-attachments/assets/749e9a10-c68b-4664-bcd6-248310100af6"> | <img width="1024" alt="has version - external_href" src="https://github.com/user-attachments/assets/9237fc77-914a-423d-be4c-e5ec12e0a17c"> |

#### No External HREF, No Version

| render | data `href` | data `external_href` |
| - | - | - |
| <img width="1024" alt="no version - render" src="https://github.com/user-attachments/assets/6f785630-407b-426f-bac2-26c6416a1669"> | <img width="1024" alt="no version - href" src="https://github.com/user-attachments/assets/b91f0330-9a0f-4ef7-ac58-6b02f469eae8"> | <img width="1024" alt="no version - external_href" src="https://github.com/user-attachments/assets/6b3bf4cb-59ce-401b-9730-4cd307839e38"> |

</details>

<details><summary>

### Fieldset Change

</summary>

<img width="653" alt="Screenshot 2024-08-16 at 4 44 14 PM" src="https://github.com/user-attachments/assets/1dde6137-28cf-4368-9eb8-d4ae38c8c215">

</details>
